src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,122 @@
 package core.basesyntax;
+import java.util.Arrays;
+import java.util.NoSuchElementException;
 
 public class ArrayList<T> implements List<T> {
+    private Object[] arrayList;
+    private int size;
+    private final int defaultSize = 10;
+
+    public ArrayList() {
+        arrayList = new Object[defaultSize];
+        size = 0;
+    }
+
+    public void ensureCapacity() {
+        if (size == arrayList.length) {
+            arrayList = Arrays.copyOf(arrayList, arrayList.length + (arrayList.length >> 1));
+        }
+    }
+
+    public void ensureCapacity(int requiredCapacity) {
+        if (requiredCapacity > arrayList.length) {
+            int newCapacity = Math.max((int) (arrayList.length * 1.5), requiredCapacity);
+            arrayList = Arrays.copyOf(arrayList, newCapacity);
+        }
+    }
+
+    public void ensureIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+    }
+
+    public void ensureIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
```

Comment:
Ensure the `ArrayListIndexOutOfBoundsException` class is implemented to handle invalid index cases .